### ISSUE TITLE
default IPv6 not working for socket starting in nccl on macOS

### ISSUE THREAD

**Issue by llv22** [llv22/nccl-osx#2]
How to disable/enable IPv6, refer to https://vladtalks.tech/vpn/how-to-disable-ipv6-on-mac
This issue may be relevant with https://github.com/llv22/nccl-osx/issues/1

```bash
(base) Orlando:nccl-tests-macOS-cuda llv23$ NCCL_DEBUG=INFO NCCL_SOCKET_FAMILY=AF_INET6 ./build/all_reduce_perf -b 8 -e 128M -f 2 -g 2
# nThread 1 nGpus 2 minBytes 8 maxBytes 134217728 step: 2(factor) warmup iters: 5 iters: 20 validation: 1 
#
# Using devices
#   Rank  0 Pid  79306 on    Orlando device  0 [0x45] GeForce GTX 1080
#   Rank  1 Pid  79306 on    Orlando device  1 [0xc4] GeForce GTX 1080
Orlando:79306:439418 [0] NCCL INFO Bootstrap : Using [0]en5:fe80::aede:48ff:fe00:1122%en5<0> [1]en0:fe80::1424:144b:157c:d2ce%en0<0> [2]awdl0:fe80::78b7:2cff:fe01:b866%awdl0<0> [3]utun0:fe80::c621:5147:9023:ea26%utun0<0> [4]utun2:fe80::639b:4048:43d:6a7b%utun2<0>
Orlando:79306:439418 [0] NCCL INFO NET/Plugin : No plugin found (libnccl-net.so), using internal implementation
Orlando:79306:439418 [0] misc/ibvwrap.cc:63 NCCL WARN Failed to open libibverbs.so[.1]
Orlando:79306:439418 [0] NCCL INFO NET/Socket : Using [0]en5:fe80::aede:48ff:fe00:1122%en5<0> [1]en0:fe80::1424:144b:157c:d2ce%en0<0> [2]awdl0:fe80::78b7:2cff:fe01:b866%awdl0<0> [3]utun0:fe80::c621:5147:9023:ea26%utun0<0> [4]utun2:fe80::639b:4048:43d:6a7b%utun2<0>
Orlando:79306:439418 [0] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61278>
Orlando:79306:439425 [0] NCCL INFO adjust filesLimit.rlim_cur to OPEN_MAX = 10240
Orlando:79306:439418 [1] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61279>
NCCL version 2.5.7+cuda10.1
Orlando:79306:439473 [0] NCCL INFO adjust filesLimit.rlim_cur to OPEN_MAX = 10240
Orlando:79306:439418 [1] NCCL INFO to create ncclGroupThreads[0] id = 439474, p = 0x70000c4e4000 to terminate
Orlando:79306:439474 [0] NCCL INFO ncclAsyncThreadMain: thread id - 439474, comm 0x7f82e5dd84e0 rank 2 nranks 0 cudaDev 0 - begin to work
Orlando:79306:439475 [0] NCCL INFO ncclAsyncThreadMain: thread id - 439475, comm 0x7f82e5dd84e8 rank 2 nranks 1 cudaDev 1 - begin to work
Orlando:79306:439418 [1] NCCL INFO to create ncclGroupThreads[1] id = 439475, p = 0x70000c567000 to terminate
Orlando:79306:439418 [1] NCCL INFO to wait ncclGroupThreads[0] id = 439474, p = 0x70000c4e4000 to terminate
Orlando:79306:439475 [1] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61280>
Orlando:79306:439475 [1] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61281>
Orlando:79306:439475 [1] NCCL INFO Connecting to socket fe80::aede:48ff:fe00:1122%en5<61279>
Orlando:79306:439475 [1] include/socket.h:405 NCCL WARN Connect to fe80::aede:48ff:fe00:1122%en5<61279> failed : Undefined error: 0, code : 0
Orlando:79306:439475 [1] NCCL INFO bootstrap.cc:96 -> 2
Orlando:79306:439475 [1] NCCL INFO bootstrap.cc:317 -> 2
Orlando:79306:439475 [1] NCCL INFO init.cc:516 -> 2
Orlando:79306:439475 [1] NCCL INFO init.cc:811 -> 2
Orlando:79306:439475 [1] NCCL INFO binding to core 0 from thread 439475
Orlando:79306:439475 [1] NCCL INFO group.cc:77 -> 2 [Async thread]
Orlando:79306:439474 [0] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61283>
Orlando:79306:439474 [0] NCCL INFO Listening on socket fe80::aede:48ff:fe00:1122%en5<61284>
Orlando:79306:439474 [0] NCCL INFO Connecting to socket fe80::aede:48ff:fe00:1122%en5<61279>
Orlando:79306:439474 [0] include/socket.h:405 NCCL WARN Connect to fe80::aede:48ff:fe00:1122%en5<61279> failed : Undefined error: 0, code : 0
Orlando:79306:439474 [0] NCCL INFO bootstrap.cc:96 -> 2
Orlando:79306:439474 [0] NCCL INFO bootstrap.cc:317 -> 2
Orlando:79306:439474 [0] NCCL INFO init.cc:516 -> 2
Orlando:79306:439474 [0] NCCL INFO init.cc:811 -> 2
Orlando:79306:439474 [0] NCCL INFO binding to core 0 from thread 439474
Orlando:79306:439474 [0] NCCL INFO group.cc:77 -> 2 [Async thread]
Orlando:79306:439418 [1] group.cc:198 NCCL WARN retrieve ncclGroupArgs[0] failed, return code: 2
Orlando:79306:439418 [1] NCCL INFO to wait ncclGroupThreads[1] id = 439475, p = 0x70000c567000 to terminate
Orlando:79306:439418 [1] group.cc:198 NCCL WARN retrieve ncclGroupArgs[1] failed, return code: 2
Orlando:79306:439418 [1] NCCL INFO init.cc:882 -> 2
Orlando: Test NCCL failure common.cu:799 'unhandled system error'
```

**Comment by llv22** [comment]
After investigation, the issue lies at socket server can't be successfully started on macOS with IPv6 address. In order to make the  NCCL-macOS-2.5.7-1 release working, I used __APPLE__ and __MACH__ macro in nccl code to disable IPv6 under macOS platform. It's better to do experiment and fix this issue separately via  https://github.com/llv22/nccl-osx/blob/dev-2.5.7-for-jax/test/ipv6_socket.cc. So we don't have to ensure the long process of launching the nccl network booting process. The hotfix is released as NCCL-macOS-2.5.7-2 release, please check in https://github.com/llv22/nccl-osx/releases/tag/nccl-macOS-2.5.7-2-release-full

**Comment by llv22** [comment]
close this issue, as fixed in https://github.com/llv22/nccl-osx/releases/tag/nccl-macOS-2.5.7-2-release-full
